### PR TITLE
Add timeout in dispatcher, AsConsumer and Seek

### PR DESCRIPTION
**Pull request by XuanYang-cn**
See also: #25309

### CODE REVIEW

**Comment by XuanYang-cn** [comment]
rerun ut

**Comment by XuanYang-cn** [comment]
rerun ut

**Comment by XuanYang-cn** [comment]
rerun ut

File: pkg/util/retry/retry.go [hunk 1]
```diff
@@ -69,6 +73,18 @@ func Do(ctx context.Context, fn func() error, opts ...Option) error {
 
 // errUnrecoverable is error instance for unrecoverable.
 var errUnrecoverable = errors.New("unrecoverable error")
+var errContextDone = errors.New("context done error")
```
Comment (congqixia) [inline pkg/util/retry/retry.go:76]: maybe using `context.DeadlineExceeded` and `context.Canceled` will do
no need to introduce new error type here?

File: pkg/util/retry/retry.go [hunk 2]
```diff
@@ -69,6 +73,18 @@ func Do(ctx context.Context, fn func() error, opts ...Option) error {
...
+
+// ContextDone method wrap an error to context done error. This will make retry
+// quick return.
+func ContextDone(err error) error {
+	return merr.Combine(err, errContextDone)
+}
+
+// IsContextDone is used to judge whether the error is wrapped by errContextDone.
+func IsContextDone(err error) bool {
```
Comment (yah01) [inline pkg/util/retry/retry.go:85]: `merr.IsCanceledOrTimeout` may be what you are finding, simply combining the errors with `ctx.Err()` should work well

**Pull request merged**